File: wisefy/src/androidTest/java/com/isupatches/wisefy/RemoveNetworkTests.java
package com.isupatches.wisefy;


import android.net.wifi.WifiConfiguration;
import com.isupatches.wisefy.callbacks.RemoveNetworkCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import org.junit.Test;
import static com.isupatches.wisefy.base.TestUtils.TEST_SSID;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.after;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class RemoveNetworkTests extends BaseAndroidJUnit4TestClass {

    @Test
    public void sync_failure_nullSSIDParam() {
        assertEquals(false, mWiseFy.removeNetwork(null));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
    }

    @Test
    public void sync_failure_missingPrerequisites() {
        missingPrerequisite();
        assertEquals(false, mWiseFy.removeNetwork(TEST_SSID));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
    }

    @Test
    public void sync_failure_noSavedNetwork() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(null);

        assertEquals(false, mWiseFy.removeNetwork(TEST_SSID));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
    }

    @Test
    public void sync_failure() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        when(mMockWiFiManager.removeNetwork(anyInt())).thenReturn(false);

        assertEquals(false, mWiseFy.removeNetwork(TEST_SSID));
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetwork(anyInt());

    }

    @Test
    public void sync_success() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        when(mMockWiFiManager.removeNetwork(anyInt())).thenReturn(true);

        assertEquals(true, mWiseFy.removeNetwork(TEST_SSID));
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetwork(anyInt());
    }

    @Test
    public void async_failure_nullSSIDParam() {
        RemoveNetworkCallbacks mockCallbacks = mock(RemoveNetworkCallbacks.class);
        mWiseFy.removeNetwork(null, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetworkWiseFyFailure(WiseFyCodes.MISSING_PARAMETER);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
    }

    @Test
    public void async_failure_nullSSIDParam_nullCallback() {
        try {
            mWiseFy.removeNetwork(null, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_missingPrerequisites() {
        missingPrerequisite();
        RemoveNetworkCallbacks mockCallbacks = mock(RemoveNetworkCallbacks.class);
        mWiseFy.removeNetwork(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetworkWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
    }

    @Test
    public void async_failure_missingPrerequisites_nullCallback() {
        missingPrerequisite();
        try {
            mWiseFy.removeNetwork(TEST_SSID, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_noSavedNetwork() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(null);

        RemoveNetworkCallbacks mockCallbacks = mock(RemoveNetworkCallbacks.class);
        mWiseFy.removeNetwork(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkNotFoundToRemove();
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
    }

    @Test
    public void async_failure_noSavedNetwork_nullCallback() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(null);

        try {
            mWiseFy.removeNetwork(TEST_SSID, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).removeNetwork(anyInt());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        when(mMockWiFiManager.removeNetwork(anyInt())).thenReturn(false);

        RemoveNetworkCallbacks mockCallbacks = mock(RemoveNetworkCallbacks.class);
        mWiseFy.removeNetwork(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureRemovingNetwork();
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetwork(anyInt());
    }

    @Test
    public void async_failure_nullCallback() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        when(mMockWiFiManager.removeNetwork(anyInt())).thenReturn(false);

        try {
            mWiseFy.removeNetwork(TEST_SSID, null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetwork(anyInt());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        when(mMockWiFiManager.removeNetwork(anyInt())).thenReturn(true);

        RemoveNetworkCallbacks mockCallbacks = mock(RemoveNetworkCallbacks.class);
        mWiseFy.removeNetwork(TEST_SSID, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkRemoved();
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetwork(anyInt());
    }

    @Test
    public void async_success_nullCallback() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        when(mMockWiFiManager.removeNetwork(anyInt())).thenReturn(true);

        try {
            mWiseFy.removeNetwork(TEST_SSID, null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).removeNetwork(anyInt());
        } catch (NullPointerException npe) {
            fail();
        }
    }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/RemoveNetworkTests.kt
package com.isupatches.wisefy

import com.isupatches.wisefy.callbacks.RemoveNetworkCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER
import com.isupatches.wisefy.internal.base.BaseInstrumentationTest
import org.junit.Assert.assertEquals
import org.junit.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to remove a network from a device's list of saved networks.
 *
 * @author Patches
 * @since 3.0
 */
internal class RemoveNetworkTests : BaseInstrumentationTest() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.removeNetwork_failure()
        assertEquals(false, wisefy.removeNetwork(TEST_SSID))
        verificationUtil.didNotTryToRemoveNetwork()
    }

    @Test fun sync_failure_noSavedNetwork() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        assertEquals(false, wisefy.removeNetwork(TEST_SSID))
        verificationUtil.didNotTryToRemoveNetwork()
    }

    @Test fun sync_failure() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockNetworkUtil.removeNetwork(false)
        assertEquals(false, wisefy.removeNetwork(TEST_SSID))
        verificationUtil.triedToRemoveNetwork()
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockNetworkUtil.removeNetwork(true)
        assertEquals(true, wisefy.removeNetwork(TEST_SSID))
        verificationUtil.triedToRemoveNetwork()
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.removeNetwork_failure()
        val mockCallbacks = mock(RemoveNetworkCallbacks::class.java)
        wisefy.removeNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
        verificationUtil.didNotTryToRemoveNetwork()
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.removeNetwork_failure()
        nullCallbackUtil.callRemoveNetwork(TEST_SSID)
        verificationUtil.didNotTryToRemoveNetwork()
    }

    @Test fun async_failure_noSavedNetwork() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        val mockCallbacks = mock(RemoveNetworkCallbacks::class.java)
        wisefy.removeNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkNotFoundToRemove()
        verificationUtil.didNotTryToRemoveNetwork()
    }

    @Test fun async_failure_noSavedNetwork_nullCallback() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        nullCallbackUtil.callRemoveNetwork(TEST_SSID)
        verificationUtil.didNotTryToRemoveNetwork()
    }

    @Test fun async_failure() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockNetworkUtil.removeNetwork(false)
        val mockCallbacks = mock(RemoveNetworkCallbacks::class.java)
        wisefy.removeNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureRemovingNetwork()
        verificationUtil.triedToRemoveNetwork()
    }

    @Test fun async_failure_nullCallback() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockNetworkUtil.removeNetwork(false)
        nullCallbackUtil.callRemoveNetwork(TEST_SSID)
        verificationUtil.triedToRemoveNetwork()
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockNetworkUtil.removeNetwork(true)
        val mockCallbacks = mock(RemoveNetworkCallbacks::class.java)
        wisefy.removeNetwork(TEST_SSID, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkRemoved()
        verificationUtil.triedToRemoveNetwork()
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.removeNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockNetworkUtil.removeNetwork(true)
        nullCallbackUtil.callRemoveNetwork(TEST_SSID)
        verificationUtil.triedToRemoveNetwork()
    }
}
